### PR TITLE
RES: Support autoderef using language attribute

### PR DESCRIPTION
**Pull request by farodin91**
I uses an idea by @kumbayo. Is it okay?

### CODE REVIEW

**Comment by matklad** [comment]
Yep, the implementation looks right! 

I think we would need to be more precise one day about autoderef behavior (current `stripAnyRefs` thing is an adhoc hack), but this can be handled later, as long as we have necessary tests in place!

But one thing we need to check is that we don't enter an infinite loop if A derefs to B and B derefs to A. Could you add a test for this as well?

**Comment by farodin91** [comment]
Will also look for lang attributes, for better correlations. I will also support `*`.

@matklad Any idea to fix the bug access tree?

**Comment by matklad** [comment]
This line causes the exception: `it.traitRef?.resolveToTrait?.identifier?.text`.

When you access the text of any node, a compact [Stub based](http://www.jetbrains.org/intellij/sdk/docs/basics/indexing_and_psi_stubs/stub_indexes.html) is switched to a more more slow PSI-based one. This test checks that resolve operates purely on stubs. 

You can see what info is available via stubs in `StubImplementations.kt`. For example, the trait's name property is stubified.

**Comment by matklad** [comment]
You can see a recent PR (https://github.com/intellij-rust/intellij-rust/pull/1156) for an example of how to add more information to stubs, should you need it.

**Comment by vlad20012** [comment]
I found a better solution that use "lang" attributes in this [branch](https://github.com/kumbayo/intellij-rust/commits/autocompletion_primitive_and_deref) and [commit](https://github.com/kumbayo/intellij-rust/commit/47f462777205b3f01327289ef4e6092fd00d7c8f)

**Comment by farodin91** [comment]
@vlad20012 I know started to implement this and i will move lang attributes to stubs.

**Comment by farodin91** [comment]
@matklad It should solve all current issues? 

My intention was to support `Vec::new().iter().next()` and so on? For this i need to add a default marco. This should be in a different PR.

**Comment by farodin91** [comment]
@matklad Any tip to clean my local history?

**Comment by farodin91** [comment]
I want also to extend the popup to show this is a autoderef function. Is this wanted and how could i implement it?

**Comment by farodin91** [comment]
@matklad What do you think here?
> I think we would need to be more precise one day about autoderef behavior (current stripAnyRefs thing is an adhoc hack), but this can be handled later, as long as we have necessary tests in place!

**Comment by vlad20012** [comment]
I think it's desirable to add several tests independent of the std to `RsTypeAwareResolveTest`

**Comment by vlad20012** [comment]
And tests for multiple deref (A to B & B to C)
And test for recursive deref (A to B & B to A)

**Comment by farodin91** [comment]
I will also add support for `deref_mut`.

File: src/main/kotlin/org/rust/lang/core/resolve/indexes/RsImplIndex.kt [hunk 1]
```diff
@@ -56,6 +72,14 @@ class RsImplIndex : AbstractStubIndex<RustTypeFingerprint, RsImplItem>() {
                 // struct S; impl<S: Tr1> Tr2 for S {}
                 ty != null && ty.javaClass == target.javaClass && ty.canUnifyWith(target, project)
             }
+            val traitDeref = collection.find { it.traitRef?.resolveToTrait?.langAttribute == "deref" }
```
Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/resolve/indexes/RsImplIndex.kt]: Yeah, the `langItems` is the right approach here! 

Let's perhaps create a `src/main/kotlin/org/rust/lang/core/resolve/LangItems.kt` file and move the lang-items related logic there? We have a ton of similar use case already, and it's high time we start consolidate it. 

For know, I think the file needs just one `fun isDerefTrait(trait: RsTraitItem): Boolean` one-liner.

**Comment by matklad** [comment]
>I will also add support for deref_mut.

We can make it in a separate PR I think

**Comment by farodin91** [comment]
> We can make it in a separate PR I think
That should be a easy one.

**Comment by vlad20012** [comment]
I found a little problem. If you resolve all impls for deref `Target` type as impls for this type, it leads to false hits to trait bound. E.g. this test fails:
```rust
struct A;
struct B;
trait Tr {}
impl Tr for B {}
#[lang = "deref"]
trait Deref { type Target; }
impl Deref for A { type Target = B; }

struct S<T>(T);
impl<T: Tr> S<T> { fn bar(&self) {} }

fn foo(a: S<A>) {
    a.bar();
    //^ unresolved
}
```
I think it's reasonable to apply deref logic not for impls resolution, but for method resolution. I.e. move your code to `findMethodsAndAssociatedFunctionsFor`.

**Comment by farodin91** [comment]
@matklad Resolved
@vlad20012 Done a bit different to use caching.

**Comment by matklad** [comment]
Let's merge this!

bors r+

I really don't like how dereferencing ends up handled directly while queering the index, will try to move it elsewhere later today, together with `stripAnyRef` and fingerprinting logic. 

Thanks!

**Pull request merged**